cmd/util/ledger/migrations/cadence_value_diff.go

```diff
@@ -0,0 +1,745 @@
+					getStorageMapKeys(newStorageMap),
+				),
+			})
+
+		return nil
+	}
+
+	if oldStorageMap != nil && newStorageMap == nil {
+		dr.reportWriter.Write(
+			difference{
+				Address: dr.address.Hex(),
+				Domain:  domain,
+				Kind:    diffKindString[storageMapExistDiffKind],
+				Msg: fmt.Sprintf(
+					"new storage map doesn't exist, old storage map has %d elements with keys %v",
+					oldStorageMap.Count(),
+					getStorageMapKeys(oldStorageMap),
+				),
+			})
+
+		return nil
+	}
+
+	oldKeys := getStorageMapKeys(oldStorageMap)
+	newKeys := getStorageMapKeys(newStorageMap)
+
+	onlyOldKeys, onlyNewKeys, sharedKeys := diff(oldKeys, newKeys)
+
+	// Log keys only present in old storage map
+	if len(onlyOldKeys) > 0 {
+		dr.reportWriter.Write(
+			difference{
+				Address: dr.address.Hex(),
+				Domain:  domain,
+				Kind:    diffKindString[storageMapKeyDiffKind],
+				Msg: fmt.Sprintf(
+					"old storage map has %d elements with keys %v, that are not present in new storge map",
+					len(onlyOldKeys),
+					onlyOldKeys,
+				),
+			})
+	}
+
+	// Log keys only present in new storage map
+	if len(onlyNewKeys) > 0 {
+		dr.reportWriter.Write(
+			difference{
+				Address: dr.address.Hex(),
+				Domain:  domain,
+				Kind:    diffKindString[storageMapKeyDiffKind],
+				Msg: fmt.Sprintf(
+					"new storage map has %d elements with keys %v, that are not present in old storge map",
+					len(onlyNewKeys),
+					onlyNewKeys,
+				),
+			})
+	}
+
+	// Compare elements present in both storage maps
+	for _, key := range sharedKeys {
+
+		trace := fmt.Sprintf("%s[%v]", domain, key)
+
+		var mapKey interpreter.StorageMapKey
+
+		switch key := key.(type) {
+		case interpreter.StringAtreeValue:
+			mapKey = interpreter.StringStorageMapKey(key)
+
+		case interpreter.StringStorageMapKey:
+			mapKey = key
+
+		case interpreter.Uint64StorageMapKey:
+			mapKey = key
+
+		default:
+			dr.reportWriter.Write(
+				diffProblem{
+					Address: dr.address.Hex(),
+					Domain:  domain,
+					Kind:    diffErrorKindString[storageMapKeyNotImplementingStorageMapKeyDiffErrorKind],
+					Trace:   trace,
+					Msg: fmt.Sprintf(
+						"invalid storage map key %v (%T), expected interpreter.StorageMapKey",
+						key,
+						key,
+					),
+				})
+			continue
+		}
+
+		oldValue := oldStorageMap.ReadValue(nopMemoryGauge, mapKey)
+
+		newValue := newStorageMap.ReadValue(nopMemoryGauge, mapKey)
+
+		hasDifference := dr.diffValues(
+			oldRuntime.Interpreter,
+			oldValue,
+			newRuntime.Interpreter,
+			newValue,
+			domain,
+			trace,
+		)
+		if hasDifference {
+			if dr.verboseLogging {
+				// Log potentially large values at top level only when verbose logging is enabled.
+				dr.reportWriter.Write(
+					difference{
+						Address:  dr.address.Hex(),
+						Domain:   domain,
+						Kind:     diffKindString[storageMapValueDiffKind],
+						Msg:      "storage map elements are different",
+						Trace:    trace,
+						OldValue: oldValue.String(),
+						NewValue: newValue.String(),
```

Comment:
We might need an extended string formatting function here. The `interpreter.Value.String` implementation often does not include the static type of the value. 

For example, dictionaries and arrays have a static type, which specifies which values may be stored in the container.

A significant portion of the Cadence 1.0 migration is migrating the static type of stored values.

We can improve this in a future PR though

Reply:
> A significant portion of the Cadence 1.0 migration is migrating the static type of stored values.

Diff logs different static types under "cadence_value_static_type_diff" kind.  It also logs trace of value, old and new static types without verbose logging enabled. 

When verbose logging isn't enabled, diff logs detailed information at lowest level.  So verbose logging is only needed if more information is required.

> We might need an extended string formatting function here. The interpreter.Value.String implementation often does not include the static type of the value.

Good point!  I pushed 4d03d8d2eb67f3571bf9199d5a7fbb0e2949b147 to add old and new static type (along with old and new values) in diff logs when verbose logging is enabled.  

If we want to include all elements' static type info, maybe it would be more efficient to do it on Cadence side.

Reply:
Nice! Sounds good 👍